### PR TITLE
Fix reset partition error when restarting a stopped replica

### PR DESCRIPTION
**Pull request by jsjtzyy**
This PR will fix two issues regarding "stop replica":

1. Previous stop replica method didn't change the state of replica in Helix state engine (only added it to "stopped" list in InstanceConfig). This becomes a problem for leader-based replication when a leader replica is stopped and Helix still believes it's leader, which will block cross-colo replication. 

2. Since replica state doesn't change in Helix while being stopped,  the stopped replica may still stay in STANDBY/LEADER state when we try to restart it. Our code calls "resetPartitionState" at the end of "start replica". This will throw exception because Helix doesn't support resetting a non-error replica. 

To address above issues, this PR makes following changes
(1)  adds logic to explicitly disable replica when stopping a replica (to trigger state transition and potential leader election);
(2) checks if the stopped replica is in ERROR state and decide whether to reset the partition.

### CODE REVIEW

**Review by justinlin-linkedin — APPROVED** [review]
LGTM

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 1]
```diff
@@ -434,10 +434,14 @@ private ServerErrorCode handleStartStoreRequest(PartitionId partitionId) {
...
     }
-    // startBlobStore should guarantee that store is started, so return value shouldn't be null
-    Store store = storeManager.getStore(partitionId);
-    // 4. reset partition state if it's offline (will trigger state transition)
-    if (store.getCurrentState() == ReplicaState.OFFLINE && clusterParticipant != null) {
+    // 4. reset partition state if it's in error state (will trigger state transition)
+    boolean isLocalStoreInErrorState =
+        partitionId.getReplicaIdsByState(ReplicaState.ERROR, currentNode.getDatacenterName())
+            .stream()
+            .map(r -> r.getDataNodeId().getHostname())
```
Comment (cgtz) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:441]: nit: can change this code slightly to not require the intermediate set:
```java
boolean isLocalStoreInErrorState =
    partitionId.getReplicaIdsByState(ReplicaState.ERROR, currentNode.getDatacenterName())
        .stream()
        .anyMatch(r -> r.getDataNodeId().getHostname().equals(currentNode.getHostname()));
```

**Review by cgtz — APPROVED** [review]
lgtm!

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java (hunk 1 above)
Comment (jsjtzyy) [reply ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:441]: Yeah, this seems better. Let me make the change.

**Comment by cgtz** [comment]
Interesting, one test got a life version conflict:
```
com.github.ambry.server.RouterServerSSLTest > interleavedOperationsTest[0] FAILED

    java.lang.Exception: Exception occured in operation: undeleteBlob-15 The blob id is AAYQAm9LK88AAQAAAAAAAAAC3kThS6TJQ6uucd2E7Sn0TA

        at com.github.ambry.server.RouterServerTestFramework$TestFuture.get(RouterServerTestFramework.java:758)

        at com.github.ambry.server.RouterServerTestFramework$11.check(RouterServerTestFramework.java:482)

        at com.github.ambry.server.RouterServerTestFramework.checkOperationChains(RouterServerTestFramework.java:132)

        at com.github.ambry.server.RouterServerSSLTest.interleavedOperationsTest(RouterServerSSLTest.java:184)

        Caused by:

        java.util.concurrent.ExecutionException: com.github.ambry.router.RouterException: LifeVersion of AAYQAm9LK88AAQAAAAAAAAAC3kThS6TJQ6uucd2E7Sn0TA from Replica Mount Path /tmp/mountpathfile620020 Replica Path /tmp/mountpathfile620020/replica620022 is different than the lifeVersion from replica Mount Path /tmp/mountpathfile620050 Replica Path /tmp/mountpathfile620050/replica620052 1 != 2 Error: LifeVersionConflict

            at com.github.ambry.router.FutureResult.resultOrThrow(FutureResult.java:113)

            at com.github.ambry.router.FutureResult.get(FutureResult.java:108)

            at com.github.ambry.server.RouterServerTestFramework$TestFuture.get(RouterServerTestFramework.java:755)

            ... 3 more

            Caused by:

            com.github.ambry.router.RouterException: LifeVersion of AAYQAm9LK88AAQAAAAAAAAAC3kThS6TJQ6uucd2E7Sn0TA from Replica Mount Path /tmp/mountpathfile620020 Replica Path /tmp/mountpathfile620020/replica620022 is different than the lifeVersion from replica Mount Path /tmp/mountpathfile620050 Replica Path /tmp/mountpathfile620050/replica620052 1 != 2 Error: LifeVersionConflict
```
https://gradle.com/s/czpbgzmpa4ol6

Going to rerun the int tests.

**Pull request merged**